pkg/controller/service/with.go

```diff
@@ -1,250 +1,541 @@
-			return
+	var configs []authz.Config
+	if superAdmin {
+		configs = []authz.Config{
+			authz.AdminAllRole,
 		}
-
-		var deviceID string
-		if strings.Contains(device, "_") {
-			deviceID = device
-		} else {
-			device, err := s.devices.LookupDevice(r.Context(), device, projectID)
-			if err == store.ErrDeviceNotFound {
-				http.Error(w, err.Error(), http.StatusNotFound)
-				return
+	} else {
+		for _, roleID := range roles {
+			role, err := s.roles.GetRole(r.Context(), roleID, project.ID)
+			if err == store.ErrRoleNotFound {
+				continue
 			} else if err != nil {
-				log.WithError(err).Error("lookup device")
+				log.WithError(err).Error("get role")
 				w.WriteHeader(http.StatusInternalServerError)
 				return
 			}
-			deviceID = device.ID
+			var config authz.Config
+			if err := yaml.Unmarshal([]byte(role.Config), &config); err != nil {
+				log.WithError(err).Error("unmarshal role config")
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			configs = append(configs, config)
 		}
+	}
 
-		handler(w, r, projectID, authenticatedUserID, authenticatedServiceAccountID, deviceID)
+	if !authz.Evaluate(requestedResource, requestedAction, configs) {
+		w.WriteHeader(http.StatusForbidden)
+		return
 	}
+
+	f(project)
 }
 
-func (s *Service) withApplicationAndDevice(handler func(http.ResponseWriter, *http.Request, string, string, string, string, string)) func(http.ResponseWriter, *http.Request, string, string, string) {
-	return func(w http.ResponseWriter, r *http.Request, projectID, authenticatedUserID, authenticatedServiceAccountID string) {
-		vars := mux.Vars(r)
-		application := vars["application"]
-		if application == "" {
-			w.WriteHeader(http.StatusBadRequest)
+func (s *Service) withDeviceAuth(w http.ResponseWriter, r *http.Request, f func(project *models.Project, device *models.Device)) {
+	vars := mux.Vars(r)
+	projectID := vars["project"]
+
+	deviceAccessKeyValue, _, _ := r.BasicAuth()
+	if deviceAccessKeyValue == "" {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	deviceAccessKey, err := s.deviceAccessKeys.ValidateDeviceAccessKey(r.Context(), projectID, hash.Hash(deviceAccessKeyValue))
+	if err == store.ErrDeviceAccessKeyNotFound {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	} else if err != nil {
+		log.WithError(err).Error("validate device access key")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	device, err := s.devices.GetDevice(r.Context(), deviceAccessKey.DeviceID, projectID)
+	if err != nil {
+		log.WithError(err).Error("get device")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	project, err := s.projects.GetProject(r.Context(), projectID)
+	if err != nil {
+		log.WithError(err).Error("get project")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	f(project, device)
+}
+
+func (s *Service) withRole(w http.ResponseWriter, r *http.Request, project *models.Project, f func(role *models.Role)) {
+	if project == nil {
```

Comment:
We could remove all these checks if we change `project *models.Project` -> `project models.Project`

Reply:
True, but we frequently pass the structs down to multiple following middleware. It's kind of like our DB functions which only return nil if they also return an err, except in this case we return an http error and simply don't call the following handler. I'd actually be in favor of removing these checks entirely, for that reason.